cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -1009,6 +1039,53 @@ struct CastFunctor<TimestampType, StringType> {
   }
 };
 
+// ----------------------------------------------------------------------
+// Binary to String
+//
+
+template <typename I>
+struct CastFunctor<
+    StringType, I,
+    typename std::enable_if<is_binary_to_string<StringType, I>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    BinaryArray binary(input.Copy());
+
+    if (options.allow_invalid_utf8) {
+      ZeroCopyData(input, output);
+      return;
+    }
+
+    // TODO(fsaintjacques): This is not the most optimal place to put this.
```

Comment:
It's probably reasonable anyway, unless we're dealing with extremely small array data.